### PR TITLE
Feature/xuc

### PR DESCRIPTION
**Pull request by orthagh**
ready to release.
I know this not testable without a xivo on your side.
But could you check if i miss some things in code (deprecation, etc)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,4 @@
 dist/
 vendor/
 .gh_token
-*.min.*
```
Comment (trasher) [inline .gitignore]: why?

File: README.md [hunk 2]
```diff
@@ -14,17 +8,18 @@ Here is the list of currently working/planned features:
...
+- [x] Click2Call (requires xivo client to handle `callto:` links)
 - [ ] Call logs
 - [ ] Directory
 
 Please contact [Teclib'](http://www.teclib-group.com) by [mail](http://www.teclib-group.com/contact/) or phone (+33 1 79 97 02 78) if you want informations for developing futures one.
 
 ## Configuration
 
-This plugin was tested with a xivo version 16.12 and should working correctly for versions above 16.04
+This plugin was tested with a xivo version 16.12 and should working correctly for versions above 16.04.
```
Comment (trasher) [inline README.md]: "should work" I guess

File: inc/apiclient.class.php [hunk 3]
```diff
@@ -67,8 +67,10 @@ function status() {
       ]);
       $line_id = is_array($line) ? end($line['items'])['id'] : false;
 
+      Toolbox::logDebug($this->auth_token);
```
Comment (trasher) [inline inc/apiclient.class.php]: Should be removed?

File: inc/line.class.php [hunk 4]
```diff
@@ -12,277 +12,152 @@ static function getTypeName($nb = 0) {
       return _n("Line", "Lines", $nb, 'xivo');
    }
 
+   function getTabNameForItem(CommonGLPI $item, $withtemplate=0) {
+      switch ($item->getType()) {
+         case "Line":
+            $nb = countElementsInTable(self::getTable(), "`lines_id` = ".$item->getID()
```
Comment (trasher) [inline inc/line.class.php]: `$nb` does not seems to be used. Also, should be conditionned

File: inc/line.class.php [hunk 5]
```diff
@@ -327,42 +260,93 @@ static function install(Migration $migration) {
...
+
+         // migrate phone_lines
+         $migration->addField('glpi_plugin_xivo_phones_lines', 'lines_id', 'integer',
+                              ['after' => 'plugin_xivo_lines_id']);
+         $migration->dropKey('glpi_plugin_xivo_phones_lines', 'unicity');
+         $migration->addKey('glpi_plugin_xivo_phones_lines', ['phones_id', 'lines_id'], 'unicity', 'UNIQUE');
+         $migration->migrationOneTable('glpi_plugin_xivo_phones_lines');
+
+         // delete preference and logs
+         $DB->query("DELETE FROM `glpi_displaypreferences`
```
Comment (trasher) [inline inc/line.class.php]: Rather use `DB::delete()` and `DB::update()` if targetting GLPI 9.3

File: inc/phone_line.class.php [hunk 6]
```diff
@@ -8,17 +8,17 @@ class PluginXivoPhone_Line extends CommonDBRelation {
...
-   static public $items_id_2 = 'plugin_xivo_lines_id';
+   static public $itemtype_2 = 'Line';
+   static public $items_id_2 = 'lines_id';
 
    function getTabNameForItem(CommonGLPI $item, $withtemplate=0) {
       switch ($item->getType()) {
-         case "PluginXivoLine":
-            $nb = countElementsInTable(self::getTable(), "`plugin_xivo_lines_id` = ".$item->getID()
+         case "Line":
+            $nb = countElementsInTable(self::getTable(), "lines_id = ".$item->getID()
```
Comment (trasher) [inline inc/phone_line.class.php]: `$nb` should be conditionned

File: inc/xuc.class.php [hunk 7]
```diff
@@ -0,0 +1,200 @@
+<?php
```
Comment (trasher) [inline inc/xuc.class.php:1]: Missing header

File: inc/xuc.class.php [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+            <div id='xuc_user_picture'>
+               <img src='".User::getThumbnailURLForPicture($picture)."'>
+            </div>
+            <div class='floating_text'>
+               <div id='xuc_fullname'></div>
+               <div id='xuc_statuses'>";
+      if ($current_config['enable_callcenter']) {
+         $out .= "<div>
+                     <label for='xuc_user_status'>".__("User", 'xivo')."</label>
+                     <select id='xuc_user_status'></select>
```
Comment (trasher) [inline inc/xuc.class.php:82]: You miss element `@name`

File: inc/xuc.class.php [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+      if (empty($caller_num)) {
+         return $data;
+      }
+
+      $r_not_digit = "[^0-9]*";
+      $regex_num = $r_not_digit.implode($r_not_digit, str_split($caller_num)).$r_not_digit;
+
+      // try to find user by its phone or mobile numbers
+      $iterator_users = $DB->request("SELECT id FROM glpi_users
```
Comment (trasher) [inline inc/xuc.class.php]: Raw query? Really? :p

File: inc/xuc.class.php [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+                                      WHERE phone  REGEXP '$regex_num'
+                                         OR mobile REGEXP '$regex_num'");
+      foreach ($iterator_users as $data_user) {
+         $userdata = getUserName($data_user["id"], 2);
+         $name     = "<b>".__("User found in GLPI:", 'xivo')."</b>".
```
Comment (trasher) [inline inc/xuc.class.php]: Why not use `printf` instead of concatenate strings?

File: inc/xuc.class.php [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+      }
+
+      // one user search for tickets
+      if (count($data['users']) > 1) {
+         // mulitple user, no redirect and return a message
+         $data['message'] = __("Multiple users found with this phone number", 'xivo');
+      } elseif (count($data['users']) == 1) {
+         $current_user     = current($data['users']);
+         $users_id         = $current_user['id'];
+         $iterator_tickets = $DB->request("SELECT glpi_tickets.id, glpi_tickets.name, glpi_tickets.content
```
Comment (trasher) [inline inc/xuc.class.php]: Replace raw query

**Review by trasher — CHANGES_REQUESTED** [review]
See my comments

File: .gitignore (hunk 1 above)
Comment (orthagh) [reply .gitignore]: i don't remember :D

Comment (cedric-anne) [reply .gitignore]: For `js/store2.min.js` ?

File: inc/xuc.class.php (hunk 8 above)
Comment (orthagh) [reply inc/xuc.class.php:82]: js form

File: inc/xuc.class.php (hunk 7 above)
Comment (orthagh) [reply inc/xuc.class.php:1]: done

File: inc/xuc.class.php [hunk 12]
```diff
@@ -169,12 +201,22 @@ function getUserInfosByPhone($params = []) {
...
+               'glpi_tickets_users' => [
+                  'FKEY' => [
+                     'glpi_tickets_users' => 'tickets_id',
+                     'glpi_tickets'       => 'id',
+                  ]
+               ]
+            ],
+            'WHERE'     => [
+               'glpi_tickets_users.type' => CommonITILActor::REQUESTER,
+               'glpi_tickets.status'     => ["<=", CommonITILObject::SOLVED],
```
Comment (trasher) [inline inc/xuc.class.php]: Was `<` not `<=` in the original query. Is that inteded?

**Review by trasher — APPROVED** [review]

File: inc/xuc.class.php (hunk 12 above)
Comment (orthagh) [reply inc/xuc.class.php]: nop

**Pull request merged**